File: 14_puzzle.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func step(polymer []rune, pair_insertion map[string]string) []rune {
	new_list := make([]rune, 0)
	polymer_str := string(polymer)
	for index, char := range polymer_str {
		new_list = append(new_list, char)
		if len(polymer_str) < index+2 {
			break
		}
		poly_substr := polymer_str[index : index+2]
		if data, ok := pair_insertion[poly_substr]; ok {
			new_list = append(new_list, rune(data[0]))
		}
	}
	return new_list
}
func count_occurence(polymer []rune) map[string]int {
	occurence := make(map[string]int)
	for _, char := range polymer {
		if data, ok := occurence[string(char)]; ok {
			occurence[string(char)] = data + 1
		} else {
			occurence[string(char)] = 1
		}
	}
	return occurence
}
func number_of_occurence(polymer []rune, pair_insertion map[string]string, step_number int) map[string]int {
	for i := 0; i < step_number; i++ {
		polymer = step(polymer, pair_insertion)
	}
	return count_occurence(polymer)
}

func merge_map(map1 map[string]int, map2 map[string]int) map[string]int {
	for k, v := range map2 {
		if _, ok := map1[k]; !ok {
			map1[k] = v
		} else {
			map1[k] += v
		}
	}
	return map1

}

func dict_of_occurence(pair_insertion map[string]string, step_number int) map[string]map[string]int {
	ans := make(map[string]map[string]int)
	for key, _ := range pair_insertion {
		ans[key] = number_of_occurence([]rune{rune(key[0]), rune(key[1])}, pair_insertion, step_number)
	}
	return ans
}
func main() {
	//file, err := os.Open("./14_easy_input.txt")
	MAX_ITERATION := 40
	BULK_ITERATION := 20
	file, err := os.Open("./14_input.txt")
	defer file.Close()
	check(err)
	scanner := bufio.NewScanner(file)
	scanner.Scan()
	polymer := strings.TrimRight(scanner.Text(), "\n")
	polymer_list := make([]rune, 0)
	for _, char := range polymer {
		polymer_list = append(polymer_list, char)
	}
	scanner.Scan() // \n
	pair_insertion := make(map[string]string)
	for scanner.Scan() {
		line := strings.Split(scanner.Text(), " -> ")
		pair_insertion[line[0]] = line[1]
	}

	for i := 0; i < BULK_ITERATION; i++ {
		polymer_list = step(polymer_list, pair_insertion)
	}
	remove_char := make(map[string]int)
	occurence := make(map[string]int)
	dict_pair_to_occurence := dict_of_occurence(pair_insertion, MAX_ITERATION-BULK_ITERATION)
	var occurence_new map[string]int
	for index := 1; index < len(polymer_list)-1; index++ {
		occurence_new = dict_pair_to_occurence[string(polymer_list[index-1:index+1])]
		char := string(polymer_list[index])
		if data, ok := remove_char[char]; ok {
			remove_char[char] = data - 1
		} else {
			remove_char[char] = -1
		}
		occurence = merge_map(occurence, occurence_new)
	}
	occurence_new = dict_pair_to_occurence[string(polymer_list[len(polymer_list)-2:])]
	occurence = merge_map(occurence, occurence_new)
	occurence = merge_map(occurence, remove_char)
	fmt.Println(occurence)
	mi := 10000000000000
	ma := 0
	for _, v := range occurence {
		if v > ma {
			ma = v
		}
		if v < mi {
			mi = v
		}
	}
	fmt.Println(ma - mi)
}

File: 2021/14_puzzle.py

from os import remove
from time import time
from typing import Dict, List

def step(polymer: List[str], pair_insertion: Dict[str, str]):
    new_list = []
    polymer = ''.join(polymer)
    i=0
    for i in range(len(polymer) - 1):
        new_list.append(polymer[i])
        poly = polymer[i:i+2]
        if poly in pair_insertion:
            new_list.append(pair_insertion[poly])
    new_list.append(polymer[i+1])
    return new_list

def number_of_occurence(polymer: List[str], pair_insertion: Dict[str, str], step_number: int, occurence: Dict[str, int]) -> Dict[str, int]:
    for _ in range(step_number):
        polymer = step(polymer, pair_insertion)
    for char in polymer:
        occurence[char] = occurence.get(char, 0) + 1
    return occurence

def memory_efficient_occurence(polymer: List[str], pair_insertion: Dict[str, str], step_count: int) -> Dict[str, int]:
    """
    dead code
    used initially to split polymer into pair and apply n step to them individually to count
    the occurence in each pair
    """
    remove_char = {}
    occurence = {}
    for index in range(1, len(polymer) - 1):
        occurence = number_of_occurence(polymer[index-1:index+1], pair_insertion, step_count, occurence)
        remove_char[polymer[index]] = remove_char.get(polymer[index], 0) + 1
    occurence = number_of_occurence(polymer[len(polymer)-2:], pair_insertion, step_count, occurence)
    for char, occu in remove_char.items():
        occurence[char] = occurence.get(char, 0) - occu
    return occurence

def dict_of_occurence(pair_insertion: Dict[str, str], step_number: int) -> Dict[str, Dict[str, int]]:
    ans = {}
    for key in pair_insertion:
        ans[key] = number_of_occurence(key, pair_insertion, step_number, {})
    return ans

if __name__ == "__main__":
    from pprint import pprint
    input = open("./14_easy_input.txt")
    input = open("./14_input.txt")
    polymer = input.readline().strip()
    polymer_list = []
    for char in polymer:
        polymer_list.append(char)
    pair_insertion = {}
    input.readline() # \n
    for line in input.readlines():
        line = line.strip().split(' -> ')
        pair_insertion[line[0]] = line[1]
    
    number_iteration = 40
    first_part = 20
    second_part = number_iteration - first_part
    for _ in range(first_part):
        polymer_list = step(polymer_list, pair_insertion)
    remove_char = {}
    occurence = {}
    dict_pair_to_occurence = dict_of_occurence(pair_insertion, second_part)
    polymer_list = ''.join(polymer_list)
    for index in range(1, len(polymer_list) - 1):
        occurence_new = dict_pair_to_occurence[polymer_list[index-1:index+1]]
        remove_char[polymer_list[index]] = remove_char.get(polymer_list[index], 0) + 1
        for char, occu in occurence_new.items():
            occurence[char] = occurence.get(char, 0) + occu
    occurence_new = dict_pair_to_occurence[polymer_list[len(polymer_list)-2:]]
    occurence = {
        char: occurence_new.get(char, 0) +
        occurence.get(char, 0) - 
        remove_char.get(char, 0)
        for char in pair_insertion.values()
    }
    mi = min(occurence.values())
    ma = max(occurence.values())
    print(ma - mi)
